### ISSUE TITLE
Cronvoy: getStatus

### ISSUE THREAD

**Issue by carloseltuerto** [envoyproxy/envoy-mobile#1519]
Cronet exposes a peculiar ```getStatus``` method. The Engine "C++ layer" is asked to return its current status, however the value is returned by invoking provided callback method. Envoy Mobile must expose a similar JNI method signature.

[JAVA](https://chromium.googlesource.com/chromium/src/+/main/components/cronet/android/api/src/org/chromium/net/UrlRequest.java#410):
```
/**
 * Queries the status of the request.
 * @param listener a {@link StatusListener} that will be invoked with
 *         the request's current status. {@code listener} will be invoked
 *         back on the {@link Executor} passed in when the request was
 *         created.
 */
public abstract void getStatus(final StatusListener listener);
```

[C++](https://chromium.googlesource.com/chromium/src/+/refs/heads/main/components/cronet/android/cronet_url_request_adapter.cc#155):
```
  void CronetURLRequestAdapter::GetStatus(
    JNIEnv* env,
    const JavaParamRef<jobject>& jcaller,            // Stream ref, an integer (similar to Envoy Mobile)
    const JavaParamRef<jobject>& jstatus_listener) { // Instance of the class with the callback method
      base::android::ScopedJavaGlobalRef<jobject> status_listener_ref;
      status_listener_ref.Reset(env, jstatus_listener);
      request_->GetStatus(
          base::BindOnce(
              &CronetURLRequestAdapter::OnStatus,
              base::Unretained(this),
              status_listener_ref));
    }
```

```getStatus``` does not return the status. Instead, the provided callback will be invoked as soon as possible, only once per ```getStatus``` invocation: this is the callback class:

```
/**
 * Listener class used with {@link #getStatus} to receive the status of a
 * {@link UrlRequest}.
 */
public abstract static class StatusListener {
  /**
   * Invoked on {@link UrlRequest}'s {@link Executor}'s thread when request
   * status is obtained.
   * @param status integer representing the status of the request. It is
   *         one of the values defined in {@link Status}.
   */
  public abstract void onStatus(int status);
}
```

And those are the possible status values. From an Envoy Mobile standpoint, those only make sense when running in "async mode".

```
/**
 * Request status values returned by {@link #getStatus}.
 */
public static class Status {
  /**
   * This state indicates that the request is completed, canceled, or is not
   * started.
   */
  public static final int INVALID = -1;

  /**
   * This state corresponds to a resource load that has either not yet begun
   * or is idle waiting for the consumer to do something to move things along
   * (e.g. when the consumer of a {@link UrlRequest} has not called
   * {@link UrlRequest#read read()} yet).
   */
  public static final int IDLE = 0;

  /**
   * When a socket pool group is below the maximum number of sockets allowed
   * per group, but a new socket cannot be created due to the per-pool socket
   * limit, this state is returned by all requests for the group waiting on an
   * idle connection, except those that may be serviced by a pending new
   * connection.
   */
  public static final int WAITING_FOR_STALLED_SOCKET_POOL = 1;

  /**
   * When a socket pool group has reached the maximum number of sockets
   * allowed per group, this state is returned for all requests that don't
   * have a socket, except those that correspond to a pending new connection.
   */
  public static final int WAITING_FOR_AVAILABLE_SOCKET = 2;

  /**
   * This state indicates that the URLRequest delegate has chosen to block
   * this request before it was sent over the network.
   */
  public static final int WAITING_FOR_DELEGATE = 3;

  /**
   * This state corresponds to a resource load that is blocked waiting for
   * access to a resource in the cache. If multiple requests are made for the
   * same resource, the first request will be responsible for writing (or
   * updating) the cache entry and the second request will be deferred until
   * the first completes. This may be done to optimize for cache reuse.
   */
  public static final int WAITING_FOR_CACHE = 4;

  /**
   * This state corresponds to a resource being blocked waiting for the
   * PAC script to be downloaded.
   */
  public static final int DOWNLOADING_PAC_FILE = 5;

  /**
   * This state corresponds to a resource load that is blocked waiting for a
   * proxy autoconfig script to return a proxy server to use.
   */
  public static final int RESOLVING_PROXY_FOR_URL = 6;

  /**
   * This state corresponds to a resource load that is blocked waiting for a
   * proxy autoconfig script to return a proxy server to use, but that proxy
   * script is busy resolving the IP address of a host.
   */
  public static final int RESOLVING_HOST_IN_PAC_FILE = 7;

  /**
   * This state indicates that we're in the process of establishing a tunnel
   * through the proxy server.
   */
  public static final int ESTABLISHING_PROXY_TUNNEL = 8;

  /**
   * This state corresponds to a resource load that is blocked waiting for a
   * host name to be resolved. This could either indicate resolution of the
   * origin server corresponding to the resource or to the host name of a
   * proxy server used to fetch the resource.
   */
  public static final int RESOLVING_HOST = 9;

  /**
   * This state corresponds to a resource load that is blocked waiting for a
   * TCP connection (or other network connection) to be established. HTTP
   * requests that reuse a keep-alive connection skip this state.
   */
  public static final int CONNECTING = 10;

  /**
   * This state corresponds to a resource load that is blocked waiting for the
   * SSL handshake to complete.
   */
  public static final int SSL_HANDSHAKE = 11;

  /**
   * This state corresponds to a resource load that is blocked waiting to
   * completely upload a request to a server. In the case of a HTTP POST
   * request, this state includes the period of time during which the message
   * body is being uploaded.
   */
  public static final int SENDING_REQUEST = 12;

  /**
   * This state corresponds to a resource load that is blocked waiting for the
   * response to a network request. In the case of a HTTP transaction, this
   * corresponds to the period after the request is sent and before all of the
   * response headers have been received.
   */
  public static final int WAITING_FOR_RESPONSE = 13;

  /**
   * This state corresponds to a resource load that is blocked waiting for a
   * read to complete. In the case of a HTTP transaction, this corresponds to
   * the period after the response headers have been received and before all
   * of the response body has been downloaded. (NOTE: This state only applies
   * for an {@link UrlRequest} while there is an outstanding
   * {@link UrlRequest#read read()} operation.)
   */
  public static final int READING_RESPONSE = 14;
}
```
